### PR TITLE
Ensure destination path exists prior to data movement

### PR DESCRIPTION
**Pull request by bdevcich**
To ensure that dcp behaves the same on each data movement, make sure the destination path exists before attempting data movement.

This also ensures that as long as the user provides a valid path with the correct permissions, dcp will have a safe place for the destination.

Path validation will need to be done upfront in nnf-sos and the nnf-dm daemon.

### CODE REVIEW

File: internal/controller/datamovement_controller.go [hunk 1]
```diff
@@ -145,6 +148,10 @@ func (r *DataMovementReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
+	statusUpdater := updater.NewStatusUpdater[*nnfv1alpha1.NnfDataMovementStatus](dm)
+	defer func() { err = statusUpdater.CloseWithStatusUpdate(ctx, r.Client.Status(), err) }()
+	// defer func() { dm.Status.SetResourceErrorAndLog(err, log) }()
```
Comment (ajfloeder) [inline internal/controller/datamovement_controller.go:153]: delete?

File: internal/controller/datamovement_controller.go [hunk 2]
```diff
@@ -516,6 +528,64 @@ func buildDMCommand(ctx context.Context, profile dmConfigProfile, hosts []string
...
+	}
+
+	// It's a file. Let Dir() determine if the destination appears to be a file or directory
+	// (trailing slash)
+	return filepath.Dir(dest), nil
+}
+
+func createDestinationDir(dest string, uid, gid uint32) error {
+	// Use exec.Command() rather than os.MkDirAll so we can set the UID/GID
+	mkdirCmd := exec.Command("mkdir", "-p", dest)
```
Comment (ajfloeder) [inline internal/controller/datamovement_controller.go:575]: You may consider something like https://github.com/NearNodeFlash/nnf-sos/blob/master/pkg/command/cmd.go.
Those Run() methods allow timeout and grab stderr and stdout in the log.

File: internal/controller/datamovement_controller.go [hunk 3]
```diff
@@ -516,6 +528,64 @@ func buildDMCommand(ctx context.Context, profile dmConfigProfile, hosts []string
...
+	mkdirCmd.SysProcAttr = &syscall.SysProcAttr{
+		Credential: &syscall.Credential{
+			Uid: uid,
+			Gid: gid,
+		},
+	}
+	if err := mkdirCmd.Run(); err != nil {
```
Comment (roehrich-hpe) [inline internal/controller/datamovement_controller.go:582]: Try `mkdirCmd.CombinedOutput()`, so you can log everything, stdout and stderr.
https://pkg.go.dev/os/exec#Cmd.CombinedOutput

File: internal/controller/datamovement_controller.go (hunk 2 above)
Comment (bdevcich) [reply internal/controller/datamovement_controller.go:575]: Good idea. I will extend that to add the ability to run as specific UID/GID.

**Review by roehrich-hpe — APPROVED** [review]

**Review by ajfloeder — APPROVED** [review]

**Review by ajfloeder — APPROVED** [review]

File: internal/controller/datamovement_controller.go (hunk 2 above)
Comment (bdevcich) [reply internal/controller/datamovement_controller.go]: PR for that is here: https://github.com/NearNodeFlash/nnf-sos/pull/277

**Comment by bdevcich** [comment]
I'd like to merge this and continue to find/fix any issues in other PRs

**Pull request merged**